reach-whitepaper.md

```diff
@@ -0,0 +1,1019 @@
+
+
+```python
+def sample_slices(
+   number_of_days,
+   cap,
+   dataset_power_law_shape,
+   n_samples
+) -> list[SliceSize]:
+"""Generates a synthetic dataset in slices (tuples of capped and uncapped numbers).
+
+
+ Args:
+   number_of_days: what time window the query is using.
+   cap: the maximal number of slices a user might contribute.
+   dataset_power_law_shape: dataset specific power-law shape param.
+   n_samples: number of slices to generate. 
+ """
+
+
+  # sample reach distribution from power-law distribution. 
+  reach_1_day = sample_discrete_power_law(
+    b=dataset_power_law_shape, 
+    n_samples=n_samples,
+    ...)
+  # approximate impact of capping.
+  # compute probability of being reported (success, not impacted by capped) 
+  # after capping.
+  p_reported = (1 - cap_discount_rate(cap=cap))
+  # for a slice of true size n, each contribution has p_reported probability to
+  # be reported. Thus after capping, observed contributions become 
+  # np.random.binomial(n, p) (there may be faster approximations) 
+  reach_1_day_capped =  np.random.binomial(reach_1_day, p_reported)
+
+
+  # approximate impact of time window
+  reach_n_days = reach_1_day * number_of_days * window_discount_rate(number_of_days)
+  reach_n_days_capped = reach_1_day_capped * number_of_days * window_discount_rate(number_of_days)
+
+
+  # round to nearest int 
+  reach_n_days = np.round(reach_n_days).astype(int)
+  reach_n_days_capped = np.round(reach_n_days_capped).astype(int)
+
+
+  return [
+    (uncapped, capped) 
+    for uncapped, capped in zip(reach_n_days, reach_n_days_capped)
+  ]
+```
+
+
+Listing 1: pseudocode to generate synthetic data. 
+
+
+<table>
+  <tr>
+   <td style="background-color: #cccccc">
+    Level
+<p>
+
+    (granularity)
+   </td>
+   <td style="background-color: #cccccc">
+    Query dimensions
+   </td>
+   <td style="background-color: #cccccc">
+    Power law shape (b) param \
+# reach per bucket
+   </td>
+  </tr>
+  <tr>
+   <td>1 (coarsest)
+   </td>
+   <td>advertiser
+   </td>
+   <td>1.01
+   </td>
+  </tr>
+  <tr>
+   <td>2
+   </td>
+   <td>advertiser x campaign
+   </td>
+   <td>1.10
+   </td>
+  </tr>
+  <tr>
+   <td>3
+   </td>
+   <td>advertiser x campaign x geo
+   </td>
+   <td>1.50
+   </td>
+  </tr>
+  <tr>
+   <td>4 (finest)
+   </td>
+   <td>advertiser x campaign x geo x ad strategy x creative
+   </td>
+   <td>1.60
+   </td>
+  </tr>
+</table>
+
+
+Table 2:  shape params used in experiments for queries with different granularity.
+
+In accordance with the pseudocode presented in Listing 1, synthetic datasets of 
+variable sizes, capping values, and time windows were generated for experimental
+evaluation.
+
+
+### Error Metrics
+
+In our experiments we are going to use $`\text{RMSRE}_\tau`$ metric defined as 
+[follows](https://developers.google.com/privacy-sandbox/relevance/attribution-reporting/design-decisions#expandable-9):
+$`
```

Comment:
Let's maybe use `$$` here to make it a block

Reply:
I tried, but apparently, GitHub doesn't render LaTeX math expressions (using $$ delimiters) correctly with the current code. Kept $ as it just works.

Reply:
I think it's meant to just use `$$` without any backticks, but that doesn't seem to be working for me either. Maybe let's use the code block syntax (using the 'language' of `math`) instead? (Second part of https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/writing-mathematical-expressions#writing-expressions-as-blocks.) That seems to be working for me.

Reply:
code block syntax works great. used it.